### PR TITLE
feat: confirma help argument

### PR DESCRIPTION
**Pull request by OpalSoPL**
### What does this PR do?

Adds  `--confirma-help` argument

### Why is this change necessary?

To add visually appealing help, available in terminal or godot

### How was this PR tested?

- manual testing
- unit tests

## Checklist:

- [x] Code is formatted correctly.
- [x] write help pages
- [x] All new and existing tests passed (unit and E2E tests).
- [x] fixme
  - [x] `TextFormatHelperTest`
  - [x] `CodeElementTest`
- [x] Documentation is updated (if necessary).
- [x] Changelog is updated (if necessary).
- [x] E2E test (probably Help Wanted)
- [x] Unit tests
  - [x] `Colors`
    - [x] `Colors.Color`
    - [x] `Colors.ColorBackground`
  - [x] `TextFormatHelper`
  - [x] `FileElementConverter`
  - [x] `FileElement` (`TextElement`)
    - [x] `LinkElement`
    - [x] `CodeElement`
    - [x] `HeaderElement`
- [x] add text formating
  - [x] bold (b)
  - [x] italic (i)
  - [x] strikethrough (s)
  - [x] underline (u)
  - [x] fill (f)
  - [x] center (c)
  - [x] link
    - [x] godot
    - [x] terminal
- [x] minimalize reset escape code recursion
- [x] add fallback for wrong page name
- [x] HeaderElement
- [x] CodeElement
- [x] LinkElement
- [x] fix: ```ERROR: This function in this node (/root/TestRunnerGame/Output) can only be accessed from either the main thread or a thread group. Use call_deferred() instead.```
- [x] internal documentation for help pages format

### CODE REVIEW

**Comment by MASSHUU12** [comment]
Have you considered using XML or JSON for the man pages? I'm not a fan of XML, but I think it would fit quite well here, and we could use attributes to customize the display of the manual. Also, C# has a built-in parser for XML and JSON. Either of these formats could be introduced later, once the functionality itself is ready.

I could also write a simple parser for Markdown (I don't want to use third-party libraries, because that would only make it harder for others to use the plugin), then the information could be retrieved directly from the docs folder. This might be an overkill, but still an interesting idea.

**Comment by OpalSoPL** [comment]
> Have you considered using XML or JSON for the man pages? I'm not a fan of XML, but I think it would fit quite well here, and we could use attributes to customize the display of the manual. Also, C# has a built-in parser for XML and JSON. Either of these formats could be introduced later, once the functionality itself is ready.

yes, just before creating this pull request, I was considering about using JSON instead of txt or Markdown as a base for manual pages 

> [...]then the information could be retrieved directly from the docs folder.

This was my original plan for this feature, but now I will stick to JSON.

**Comment by OpalSoPL** [comment]
first look of the feature
obraz
```json
{
    "version": 1,
    "data": [
        {
            "type":"header",
            "text":"Arguments",
            "level":1
        },
        {
            "type":"text",
            "text":"List of available arguments",
            "format":["i"]
        },
        {
            "type":"header",
            "text":"Legend:",
            "level":3
        },
        {
            "type":"code",
            "lines":
            [
                "[] - optional argument",
                "<> - required argument"
            ]
        },
        {
            "type":"header",
            "text":"--confirma-run",
            "level":2
        }
    ]
}
```

`fill` format is too unpredictable (why can't I get terminal width, AAAAAhhh), so I'll probably need to decide between removing the `fill` format or trying to fix it.

**Comment by MASSHUU12** [comment]
It's starting to look surprisingly good. I didn't think you would put so much work into it and expand it so much, I'm impressed.

I wonder if the `help_pages` folder should be moved from `src` to `docs`, what do you think?

Also, am I missing something? Because I created `default.json` with the contents you commented above, but Confirma doesn't display anything. I'm guessing you haven't committed the changes yet, if you need help, I'm eager to help.

Terminal screenshot

How should the `fill` behave? As far as I can see, at this point it is simply `EFormatType.fill => $“\n{text}”,`? Also, you should be able to get terminal width using `Console.WindowWidth`, and for the width of the output in the editor this should be rather easy.

**Comment by OpalSoPL** [comment]
> How should the `fill` behave? As far as I can see, at this point it is simply `EFormatType.fill => $“\n{text}”,`?

`fill` was behaving wrongly because of `Log.PrintLine` in `Help` that i forgot to remove after some testing, so it wasn't fault of `fill`
>Also, you should be able to get terminal width using `Console.WindowWidth`, and for the width of the output in the editor this should be rather easy.

For some reason, `Console.WindowWidth` is always 0. Edit: `Console.WindowWidth` is 0 in VSC debuger, my bad.

> Also, am I missing something? Because I created `default.json` with the contents you commented above, but Confirma doesn't display anything. I'm guessing you haven't committed the changes yet, if you need help, I'm eager to help.

i've fixed `ConfirmaAutoload` but never committed it to the branch
obraz

**Comment by MASSHUU12** [comment]
I wonder if it would make sense to introduce the ability to create links, just not sure about the usefulness of such a feature. Godot supports this without a problem, from what I can see, terminals also through [OSC 8](https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda#the-escape-sequence). Local files are also supported, so links could lead to, for example, detailed documentation on a topic in a Markdown file.

I don't know if such a feature is outside the scope of your PR. I would happily implement something like this.

**Comment by OpalSoPL** [comment]
My goal with this PR is to create something similar to `man` command, which can be viewed in both graphical and headless modes.

As for links. I didn't know about link in terminal. But after some experiments, I think they are amazing feature with great potential in it

**Comment by OpalSoPL** [comment]
(terminal)
obraz
(godot)
obraz

Graphical version of `--confirma-help` is still far from complete. I'll probably need to do some tweaks with theme of the `Output` node, and the `link` format isn't functional yet.

**Comment by MASSHUU12** [comment]
image

> I knew it XD

It looks super and works nicely not counting the following errors, but you probably know about them (although I wouldn't be surprised if they were some problems specific to my Godot):

image

```txt
[...]
WARNING: 8 RIDs of type "CanvasItem" were leaked.
     at: _free_rids (servers/rendering/renderer_canvas_cull.cpp:2513)
ERROR: 1 RID allocations of type 'PN13RendererDummy14TextureStorage12DummyTextureE' were leaked at exit.
ERROR: 3 RID allocations of type 'PN18TextServerAdvanced22ShapedTextDataAdvancedE' were leaked at exit.
ERROR: 1 RID allocations of type 'PN18TextServerAdvanced12FontAdvancedE' were leaked at exit.
WARNING: ObjectDB instances leaked at exit (run with --verbose for details).
     at: cleanup (core/object/object.cpp:2342)
Leaked instance: FontVariation:9223372057004212861 - Resource path:
Leaked instance: FontVariation:9223372056987435663 - Resource path:
Leaked instance: FontVariation:9223372056970658448 - Resource path:
Leaked instance: FontFile:9223372056953881388 - Resource path:
Leaked instance: StyleBoxFlat:9223372060980413697 - Resource path:
Leaked instance: StyleBoxFlat:9223372061248849169 - Resource path:
Leaked instance: StyleBoxFlat:9223372061265626386 - Resource path:
Leaked instance: ImageTexture:9223372061299180820 - Resource path:
Leaked instance: StyleBoxFlat:9223372061332735254 - Resource path:
Leaked instance: StyleBoxFlat:9223372061349512471 - Resource path:
Leaked instance: StyleBoxFlat:9223372061366289688 - Resource path:
Leaked instance: StyleBoxFlat:9223372061383066905 - Resource path:
Leaked instance: StyleBoxEmpty:9223372062289036623 - Resource path:
ERROR: Cannot get path of node as it is not in a scene tree.
   at: get_path (scene/main/node.cpp:2327)
Leaked instance: Control:26390562152 - Node path:
Leaked instance: Object:26407339369
Leaked instance: Image:9223372063564105066 - Resource path:
Leaked instance: ImageTexture:9223372063664768363 - Resource path:
Leaked instance: StyleBoxFlat:9223372063161451886 - Resource path: res://scenes/main/Main.tscn::StyleBoxFlat_m0bg1
ERROR: Cannot get path of node as it is not in a scene tree.
   at: get_path (scene/main/node.cpp:2327)
Leaked instance: ColorRect:26424116591 - Node path:
Leaked instance: Object:26440893808
ERROR: Cannot get path of node as it is not in a scene tree.
   at: get_path (scene/main/node.cpp:2327)
Leaked instance: CenterContainer:26457671025 - Node path:
Leaked instance: Object:26474448242
ERROR: Cannot get path of node as it is not in a scene tree.
   at: get_path (scene/main/node.cpp:2327)
Leaked instance: PanelContainer:26491225459 - Node path:
Leaked instance: Object:26508002676
ERROR: Cannot get path of node as it is not in a scene tree.
   at: get_path (scene/main/node.cpp:2327)
Leaked instance: MarginContainer:26524779893 - Node path:
Leaked instance: Object:26541557110
ERROR: Cannot get path of node as it is not in a scene tree.
   at: get_path (scene/main/node.cpp:2327)
Leaked instance: VBoxContainer:26558334327 - Node path:
Leaked instance: Object:26575111544
ERROR: Cannot get path of node as it is not in a scene tree.
   at: get_path (scene/main/node.cpp:2327)
Leaked instance: RichTextLabel:26591888761 - Node path:
Leaked instance: Object:26608665978
Leaked instance: TextParagraph:9223372063580882299
ERROR: Cannot get path of node as it is not in a scene tree.
   at: get_path (scene/main/node.cpp:2327)
Leaked instance: VScrollBar:26642220412 - Node path:
Leaked instance: Object:26658997629
Leaked instance: Image:9223372063681545600 - Resource path:
Hint: Leaked instances typically happen when nodes are removed from the scene tree (with `remove_child()`) but not freed (with `free()` or `queue_free()`).
ERROR: 1 resources still in use at exit.
   at: clear (core/io/resource.cpp:602)
Resource still in use: res://scenes/main/Main.tscn::StyleBoxFlat_m0bg1 (StyleBoxFlat)
[...]
```

**Comment by OpalSoPL** [comment]
> [...]not counting the following errors, but you probably know about them (although I wouldn't be surprised if they were some problems specific to my Godot):

Weird, I didn't get any error, even with `--verbose`.
Have you tried removing `.godot` directory?

File: addons/confirma/src/scenes/confirma_autoload/ConfirmaAutoload.cs [hunk 1]
```diff
@@ -57,6 +57,15 @@ private bool CheckArguments()
                 _usedConfirmaApi = true;
             }
 
+            if (arg.StartsWith(prefix + "help", InvariantCulture))
+            {
+                _ = GetTree().CallDeferred("change_scene_to_file", "uid://tx5xyqboqit7");
```
Comment (OpalSoPL) [inline addons/confirma/src/scenes/confirma_autoload/ConfirmaAutoload.cs:62]: This is causing the issue, but why?
Same approach is also used in line 237 but that isn't causing any error.

Comment (MASSHUU12) [reply addons/confirma/src/scenes/confirma_autoload/ConfirmaAutoload.cs:62]: Changing `return true;` to `return false;` seems to fix the issue.

Comment (OpalSoPL) [reply addons/confirma/src/scenes/confirma_autoload/ConfirmaAutoload.cs]: ok, this should be fixed in bb9ffbeb

**Comment by OpalSoPL** [comment]
obraz
It looks like that the `--confirma-run` has the same issue after removing `.godot` directory

**Comment by OpalSoPL** [comment]
Probably `.godot/uid_cache.bin` is a culprit of these issues (both `--confirma-run` and `--confirma-help`)
You probably had outdated `uid_cache.bin` from other branch when you ran it first time

**Comment by OpalSoPL** [comment]
@MASSHUU12 
Do you have any idea what i could do to fix `fill` in graphical mode?
What I considered to do, was using a monospaced font. Probably that's the best option to do this, but i might be wrong.

**Comment by MASSHUU12** [comment]
> @MASSHUU12 Do you have any idea what i could do to fix `fill` in graphical mode? What I considered to do, was using a monospaced font. Probably that's the best option to do this, but i might be wrong.

I think I was able to find a rather good approach, but it still needs some work. This is the result at the moment:

image

Currently, the code looks like this:

```cs
    #region Fill
    public static string Fill<T>(T text, int width = 0)
    where T : IConvertible
    {
        return Log.IsHeadless
            ? FillToTerminal(text, width)
            : FillToGodot(text, width);
    }

    // ...

    public static string FillToGodot<T>(T text, int width = 0)
    where T : IConvertible
    {
        // Can be improved by caching the widths of characters.

        string t = text?.ToString() ?? string.Empty;

        Font? font = Log.RichOutput?.GetThemeDefaultFont();
        int fontSize = Log.RichOutput?.GetThemeDefaultFontSize() ?? 16;

        if (font is null)
        {
            return t;
        }

        int windowWidth = (int)(Log.RichOutput?.GetRect().Size.X ?? width);

        int currentWidth = 0;
        foreach (char c in t)
        {
            int charWidth = (int)font.GetStringSize(c.ToString(), fontSize: fontSize).X;
            currentWidth += charWidth;
        }

        while (currentWidth < windowWidth)
        {
            t += " ";
            currentWidth += (int)font.GetStringSize(" ", fontSize: fontSize).X;
        }

        return t;
    }
    #endregion Fill
```

**Comment by OpalSoPL** [comment]
obraz
> Output of your second code after disabling word wrapping

I need to look at different formats and see how they're behaving with `fill`.

**Comment by OpalSoPL** [comment]
> I wonder if the `help_pages` folder should be moved from `src` to `docs`, what do you think?

Maybe to `docs/help_pages` instead of `docs`, so that they don't mix with those in Markdown.

PS: I forgot to address that part of comment

**Comment by OpalSoPL** [comment]
obraz
> default help page

**Comment by MASSHUU12** [comment]
That looks really nice.

> E2E test (probably Help Wanted)

I noticed that you will probably need help with E2E testing. I am not an expert in these tests, plus the technology I chose is probably not the best. I didn't want to use C# for this task, so I was looking for something that was faster and simpler to write. I tried with several different technologies, but in the end I chose TypeScript (Bun). If you have any questions, I'm happy to help. You can also email me with your Signal or Discord account name if you want a faster and more convenient conversation.

**Review by MASSHUU12 — APPROVED** [review]
LGTM

**Pull request merged**